File: fullstacks/spring-mvc/src/main/java/com/hellotimepro/springmvc/repository/FavoriteRepository.java
package com.hellotimepro.springmvc.repository;

import com.hellotimepro.springmvc.domain.FavoriteEntity;
import com.hellotimepro.springmvc.domain.FavoriteId;
import java.util.Collection;
import java.util.List;
import java.util.UUID;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

public interface FavoriteRepository extends JpaRepository<FavoriteEntity, FavoriteId> {
  List<FavoriteEntity> findByIdUserIdOrderByCreatedAtDesc(UUID userId);
  Page<FavoriteEntity> findByIdUserIdOrderByCreatedAtDesc(UUID userId, Pageable pageable);
  List<FavoriteEntity> findByIdUserIdAndIdCapsuleIdIn(UUID userId, Collection<UUID> capsuleIds);
  void deleteByIdCapsuleId(UUID capsuleId);
}

File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/repository/FavoriteRepository.kt
package com.hellotimepro.ktor.repository

import com.hellotimepro.ktor.db.Favorites
import com.hellotimepro.ktor.domain.Favorite
import org.jetbrains.exposed.sql.*
import java.util.UUID

class FavoriteRepository {
    fun find(userId: UUID, capsuleId: UUID): Favorite? =
        Favorites.select { (Favorites.userId eq userId) and (Favorites.capsuleId eq capsuleId) }
            .firstOrNull()?.toFavorite()

    fun exists(userId: UUID, capsuleId: UUID): Boolean =
        Favorites.select { (Favorites.userId eq userId) and (Favorites.capsuleId eq capsuleId) }
            .count() > 0L

    fun insert(favorite: Favorite) {
        Favorites.insert {
            it[userId] = favorite.userId
            it[capsuleId] = favorite.capsuleId
            it[createdAt] = favorite.createdAt
        }
    }

    fun delete(userId: UUID, capsuleId: UUID) {
        val cond = Op.build { (Favorites.userId eq userId) and (Favorites.capsuleId eq capsuleId) }
        Favorites.deleteWhere { cond }
    }

    fun deleteByCapsule(capsuleId: UUID) {
        val cond = Op.build { Favorites.capsuleId eq capsuleId }
        Favorites.deleteWhere { cond }
    }

    fun findByUserPaged(userId: UUID, offset: Long, limit: Int): List<Favorite> =
        Favorites.select { Favorites.userId eq userId }
            .orderBy(Favorites.createdAt to SortOrder.DESC)
            .limit(limit, offset)
            .map { it.toFavorite() }

    fun countByUser(userId: UUID): Long =
        Favorites.select { Favorites.userId eq userId }.count()

    fun findByUserAndCapsuleIds(userId: UUID, capsuleIds: Collection<UUID>): List<Favorite> =
        if (capsuleIds.isEmpty()) emptyList()
        else Favorites.select { (Favorites.userId eq userId) and (Favorites.capsuleId inList capsuleIds) }
            .map { it.toFavorite() }
}
